### ISSUE TITLE
is anything error in zero web page?

### ISSUE THREAD

**Issue by l1t1** [gcp/leela-zero#368]
First 3 lines are same vs

Matches (100 Most Recent) | Wins / Losses | Games
-- | -- | --
d5b7aec3 VS 5bcb392e | 0 : 0 | 0 / 400
d5b7aec3 VS 5bcb392e | 0 : 0 | 0 / 400
d5b7aec3 VS 5bcb392e | 20 : 51 (28.17%) | 71 / 400
5bcb392e VS fe7a6c28 | 337 : 64 (84.04%) | 401 / 400

**Comment by l1t1** [comment]
I think it should read from http://zero.sjeng.org/networks/

Name|Size|Modified
-- | -- | --
4701bb9fe03a8bdaedbe82b9872f5ba2f1fd15e3a1244db7e2ead672a873a752.gz|2798724|2017-12-13 06:21:31
bd08cfd438be41cf6038a1a96ea2e320609b7050a5917cbd00317935d0764bd6.gz|2798900|2017-12-13 06:21:23
d5b7aec30633d495bf1908a5ead955b02553188f63d56fd6d6ac13b5d0ee4232.gz|2800707|2017-12-13 06:21:08

**Comment by BHydden** [comment]
I believe this was intentional so that we can confirm that training the network should get the similar results (repeatable)

**Comment by isty2e** [comment]
@BHydden Well if you click on the links you'll notice that they're broken. I bet this is some kind of a mistake or an error.

**Comment by BHydden** [comment]
huh... it made sense to me that that was how they were testing their training process but you are right the top two have broken links currently that's very odd

**Comment by Uizmp** [comment]
They pasted the hash code but delete the first 12 chars (d5b7aec30633)

**Comment by l1t1** [comment]
d5b7aec3 VS 5bcb392e | 85 : 219 (27.96%) | 304 / 400
-- | -- | --

**Comment by isty2e** [comment]
Please state your opinion. One does not simply show something that everyone can see and expect a productive discussion.

**Comment by kimitsu** [comment]
Well, now I'm getting this
```
D:\Files\Go\leela-zero-next>autogtp.exe
AutoGTP v10
Using 1 thread(s).
curl.exe -s -J http://zero-test.sjeng.org/get-task/7
{
    "black_hash": "d5b7aec30633bd08cfd438be41cf6038a1a96ea2e320609b7050a5917cbd00317935d0764bd6",
    "cmd": "match",
    "leelaz_version": "0.9",
    "options": {
        "noise": "false",
        "playouts": "1600",
        "randomcnt": "0",
        "resignation_percent": "1"
    },
    "options_hash": "321436",
    "random_seed": "12158812241147532830",
    "required_client_version": "10",
    "white_hash": "5bcb392e581915631ab7dd43c08ce7aae9ab51e2ff6c81793e54e69d55b6073f"
}

Required client version: 10
 -p 1600  -r 1  -m 0  -t 1  -d   --noponder  -s 12158812241147532830
curl.exe -s -O -J -w %{filename_effective} http://zero.sjeng.org/networks/d5b7aec30633bd08cfd438be41cf6038a1a96ea2e320609b7050a5917cbd00317935d0764bd6.gz
Curl filename: d5b7aec30633bd08cfd438be41cf6038a1a96ea2e320609b7050a5917cbd00317935d0764bd6.gz

gzip: d5b7aec30633bd08cfd438be41cf6038a1a96ea2e320609b7050a5917cbd00317935d0764bd6.gz: not in gzip format
Net filename: d5b7aec30633bd08cfd438be41cf6038a1a96ea2e320609b7050a5917cbd00317935d0764bd6
```

**Comment by Mardak** [comment]
@roy7, looks like there was a typo prepending `d5b7aec30633` to the actual network file.

$ curl -I http://zero.sjeng.org/networks/d5b7aec30633bd08cfd438be41cf6038a1a96ea2e320609b7050a5917cbd00317935d0764bd6.gz
HTTP/1.1 404 Not Found
$ curl -I http://zero.sjeng.org/networks/bd08cfd438be41cf6038a1a96ea2e320609b7050a5917cbd00317935d0764bd6.gz
HTTP/1.1 200 OK

As @kimitsu just pointed out, there are now /next requests with: `    "white_hash": "d5b7aec30633bd08cfd438be41cf6038a1a96ea2e320609b7050a5917cbd00317935d0764bd6"` resulting in the client stopping after:
```
curl -s -O -J -w %{filename_effective} http://zero.sjeng.org/networks/d5b7aec30633bd08cfd438be41cf6038a1a96ea2e320609b7050a5917cbd00317935d0764bd6.gz
Curl filename: d5b7aec30633bd08cfd438be41cf6038a1a96ea2e320609b7050a5917cbd00317935d0764bd6.gz
Net filename: d5b7aec30633bd08cfd438be41cf6038a1a96ea2e320609b7050a5917cbd00317935d0764bd6
```

**Comment by kimitsu** [comment]
@l1t1 /master branch only generates self-plays now, so when the new best network is updated, it will switch to that, /next branch has new network testing matches implemented, it will follow server's commands to set up matches between the current best network and the next candidate network on the list and send results to the server.

**Comment by marcocalignano** [comment]
I (apparently) did request all the 'bad' matches queued in the server and now my autogtp is playing self -played games again.

**Comment by liverex** [comment]
Should it be:

4701bb9f VS 5bcb392e 
bd08cfd4 VS 5bcb392e 
d5b7aec3 VS 5bcb392e
?

**Comment by odeint** [comment]
Yes it should be, and those networks are trained with 24000, 44000, and 84000 steps. We started testing with the 24000 step one.

**Comment by killerducky** [comment]
@roy7 in addition to the typo in the match request, I think there is some other bug. How could this match get 1 result if the name of the network was wrong?
`12/13/2017, 6:24:00 AM	d5b7aec3 VS 5bcb392e	1 : 0 (100.00%)	1 / 400`

**Comment by isty2e** [comment]
@killerducky Someone could have figure out what happened and downloaded the network and renamed it. I tried the same thing, but it was after I got only selfplay queries.

**Comment by evanroberts85** [comment]
So the 12k steps network yesterday scored only 30% and the 24k steps network scored only 26%. This suggests that close to 20k steps is the sweet spot.

**Comment by Dorus** [comment]
You make tons of assumptions there. For yesterdays network that was probably true (but the 20,001 or 19999 network might have been stronger too!).

There is a lot of randomness in training. The overall trend will be upwards, but a single learning step can go in any direction. I would be greatly surprised if 20k step is the sweet spot every time.

**Comment by evanroberts85** [comment]
That is a fair point, may be worth running this experiment every week,  maybe with -2k, current and +2k step networks and see if there is any change to what works best.

**Comment by kimitsu** [comment]
@evanroberts85 Actually, this 44k steps network is a beast! I've run a couple simulations against the reigning 5bcb392e. My jaw almost dropped after the first 4 moves by this new network playing white here http://eidogo.com/#A18g8x7I

**Comment by roy7** [comment]
Yes, the matches were entered with a wrong hash. I've fixed that in the match database so things should be back to normal now. Glad to see that the failure was handled by fallback to selfplay as we'd hope/expect. :)

**Comment by isty2e** [comment]
@roy7 But autogtp stopped after the failed download, maybe that would be the next improvement to be made.

**Comment by zediir** [comment]
It fell back to selfplay after someone had requested all 400 matches.

**Comment by roy7** [comment]
@isty2e Are you on release or /next? Maybe only /next has the error handling and fallback to selfplay in place.

@zediir What I mean is the client who gets a bad match command didn't stop running from the error.

**Comment by isty2e** [comment]
@roy7 I'm on next, the json requires client version to be 10.

**Comment by roy7** [comment]
Then I guess the error handling didn't work for you. :( Others reported autogtp fell back to selfplay, and now that matches are fixed and server restarted, you should start getting matches again. If autogtp crashed out maybe you can report it as an issue or provide any details to the existing issue or PR related to handling errors?

**Comment by marcocalignano** [comment]
@roy7  autogtp from /next exit with an error if the network cannot be found. I had to start autogtp ~800 times before I got a selfplay json that I could process since then everyone else that tryed got also a selfplay game

**Comment by isty2e** [comment]
Yeah I could get autogtp functioning only after I started to receive self-play queries.

**Comment by roy7** [comment]
Ah okay. So that's a feature request then, in the case a network is missing for a match, fall back cleanly to selfplay instead using prior best network or using the old /best-network-hash endpoint, or maybe just request a new get-task over and over until the bad match requests have flushed the system?

I'll try add something on our end so the server won't accept match requests for non-existent networks as well. Maybe that's enough and autogtp doesn't need to handle it.

**Comment by Dorus** [comment]
Be careful when you re-request after a bad match has been received. Better add some (incremental) back off feature else you start hammering the server really hard with 250+ eager clients we have soon :)

This time re-requesting solved the issue, but that might not always be the case.

**Comment by isty2e** [comment]
A sanity check on the server side would be fine IMHO.

**Comment by Ttl** [comment]
@marcocalignano  @roy7 I can fix the clients quitting. Falling back to self-play is difficult, but making it get a new task from server should be simple.

**Comment by roy7** [comment]
@ttl An exponential backoff would be appreciated. In case we have some unknown issue in the future.

**Comment by roy7** [comment]
Right now if you are self play and server crashes, you'll keep doing selfplays. I wonder what happens if you are doing a match and server crashes and you have no self play information. It should in theory try to reconnect every so often until server is back up, would be nice. :)

**Comment by Ttl** [comment]
Offline self-play was not merged. If the server crashes all clients (master and next) keep requesting work from server with exponential backoff. I'll add invalid network from server to trigger the same retry logic.

**Comment by evanroberts85** [comment]
maybe have a /get_selfplay_game call to fall back on for a single game if
/get_task fails?

**Comment by marcocalignano** [comment]
@roy7 

>  if you are self play and server crashes, you'll keep doing selfplays

not in the /next branch. Autogtp need always an  input from the server to do some work

**Comment by roy7** [comment]
@evanroberts85 Can use old /best-network-hash for that like /master clients do now. But I should probably make a new endpoint with the full JSON task info.

**Comment by Ttl** [comment]
@evanroberts85 too much work for very minor function in my opinion, at least I won't be implementing it. If some sanity checks are added on the server this failure mode shouldn't happen again.

**Comment by marcocalignano** [comment]
@Ttl  I have an Idea how to do the fallbck better (maybe) do it better. we can discuss that and then we can implement it better. So instead of fixing the client crashing try to build a function (for now) that can return few error code like 
-server not reachable
-network file corrupted
-no network present

**Comment by marcocalignano** [comment]
@roy7 what I'd like to have is a fallback json object in the match json so I know what to do even in a match situation if the server is down.

**Comment by isty2e** [comment]
IMHO a sanity check on the server and hash check on the client will be enough, except for the network disconnection. Not every functionality needs to be implemented in the client especially when it is likely to be complicated, I assume.

**Comment by Dorus** [comment]
Tbh most errors you can get while requesting games will be the result of random outages. Even if it's a server config problem, you can still just make a new request once the config has been fixed.

If anything, it is really the server that should fall back to self play (or the next evaluation match), if too many matches come back in failure.

Errors codes from the clients are nice, that's certainly something that is useful to add so the server can log what kind of things go wrong.

Just keep it simple, implement exponential backoff as @roy7 requested, and optionally add error codes.

**Comment by evanroberts85** [comment]
Doris makes a good point, ideally the server should detect if it is not getting any results back and halt that match, moving on to the next. If one client has a problem it is likely they all will.

**Comment by gcp** [comment]
This was an error on my side. I queue the matches semi-manually for now, and as I'm traveling I was doing this from a remote console on a mac. Apparently it has a bug and although it showed the correct command for queuing the match, there was actually some stuff prepanded.

**Comment by marcocalignano** [comment]
@roy7 I have a coomit in my local git that create a fallback from the last successfull  Self-played game that autogtp did so if there is a problem on the server or on the connection I can still do work till I try next time. 
Do you have an Idea how we can test it? Maybe a temp URL that I can use in my local version where you can induce errors?

**Comment by roy7** [comment]
Unplug your Ethernet cable? :)

**Comment by marcocalignano** [comment]
Ok next time I get a self-played game I will try to "climb" under my table and unplug my Ethernet cable. But this does not test all the scenario. But I guess is good enogh.

**Comment by BHydden** [comment]
@marcocalignano if the physical cable is too hard to yank you can just temporarily disable the network card
